pkg/services/live/features/testdata.go

```diff
@@ -0,0 +1,124 @@
+package features
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/centrifugal/centrifuge"
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/logger"
+	"github.com/grafana/grafana/pkg/models"
+)
+
+// TestdataRunner manages all the `grafana/dashboard/*` channels
+type testdataRunner struct {
+	publisher   models.ChannelPublisher
+	running     bool
+	speedMillis int
+	dropPercent float64
+	channel     string
+}
+
+// TestdataSupplier manages all the `grafana/testdata/*` channels
+type TestdataSupplier struct {
+	publisher models.ChannelPublisher
+}
+
+// CreateTestdataSupplier Initialize a dashboard handler
+func CreateTestdataSupplier(p models.ChannelPublisher) TestdataSupplier {
+	return TestdataSupplier{
+		publisher: p,
+	}
+}
+
+// GetHandlerForPath called on init
+func (g *TestdataSupplier) GetHandlerForPath(path string) (models.ChannelHandler, error) {
+	channel := "grafana/testdata/" + path
+
+	if path == "random-2s-stream" {
+		return &testdataRunner{
+			publisher:   g.publisher,
+			running:     false,
+			speedMillis: 2000,
+			dropPercent: 0,
+			channel:     channel,
+		}, nil
+	}
+
+	if path == "random-flakey-stream" {
+		return &testdataRunner{
+			publisher:   g.publisher,
+			running:     false,
+			speedMillis: 400,
+			dropPercent: .6,
+			channel:     channel,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("unknown channel")
+}
+
+// GetChannelOptions called fast and often
+func (g *testdataRunner) GetChannelOptions(id string) centrifuge.ChannelOptions {
+	return centrifuge.ChannelOptions{}
+}
+
+// OnSubscribe for now allows anyone to subscribe to any dashboard
+func (g *testdataRunner) OnSubscribe(c *centrifuge.Client, e centrifuge.SubscribeEvent) error {
+	if !g.running {
+		g.running = true
+
+		// Run in the background
+		go g.runRandomCSV()
```

Comment:
NOTE:  this now starts a random walk on the *first* subscribe.  The stuff in master starts it on startup.

We still need to add hooks to stop it when everyone has unsubscribed, but that will need to come in a follow on issue

Reply:
Is this threadsafe? For example if nothing is subscribed to yet, but then two or more clients subscribe at around the same time is there any guarantee that multiple threads running `runRandomCSV` won't be started?

Reply:
:crossed_fingers: I think so -- only a single handler will be initialized once, and should be blocked by this init logic:
https://github.com/grafana/grafana/pull/27583/files#diff-a2dd8b619e1ed03a64a4ad4cc0cc8e04R227